### DISCUSSION TITLE
There is no boxing conversion from 'SixLabors.ImageSharp.PixelFormats.Rgba32' to '?'.

### DISCUSSION THREAD

**Discussion by LifeguardConcepts** [SixLabors/ImageSharp#1656] [Q&A] [answered]
Hello,
I'm having a build issue which is stumping me to no end.
This is not occurring on my local dev box only on azure during our CI build but I'm at a loss as its not a typical unboxing message related as it doesn't seem to be able to identify the converting to format, hence the '?' value.

The code below is taking some direction from the ZXing bar code library which I would use his extension method directly but it has not been updated for the current release of ImageSharp.  Essentially we are using the BitMatrix to apply Pixel level changes to the image to generate a Image from a PDF417 barcode.

Error: 
Error CS0648: '' is a type not supported by the language
Error CS0315: The type 'SixLabors.ImageSharp.PixelFormats.Rgba32' cannot be used as type parameter 'TPixel' in the generic type or method 'Image<TPixel>'. There is no boxing conversion from 'SixLabors.ImageSharp.PixelFormats.Rgba32' to '?'.

Related Code:

```
private static Image<SixLabors.ImageSharp.PixelFormats.Rgba32> Render(BitMatrix matrix, ZXing.BarcodeFormat format, string content, EncodingOptions options)
{       
      var width = matrix.Width;
      var height = matrix.Height;
      var black = new Rgba32(0xFF000000);
      var white = new Rgba32(0xFFFFFFFF);

      var pixelsize = 1;

      var result = new Image<SixLabors.ImageSharp.PixelFormats.Rgba32>(width, height);
      for (int y = 0; y < matrix.Height; y++)
      {
          for (var pixelsizeHeight = 0; pixelsizeHeight < pixelsize; pixelsizeHeight++)
          {
              var rowOffset = pixelsize * y + pixelsizeHeight;

              for (var x = 0; x < matrix.Width; x++)
              {
                  var color = matrix[x, y] ? black : white;
                  for (var pixelsizeWidth = 0; pixelsizeWidth < pixelsize; pixelsizeWidth++)
                  {
                      var pixel = new Rgba32();
                      pixel.FromRgba32(color);
                      result[pixelsize * x + pixelsizeWidth, rowOffset] = pixel;
                  }
              }
              for (var x = pixelsize * matrix.Width; x < width; x++)
              {
                  var pixel = new Rgba32();
                  pixel.FromRgba32(white);
                  result[x, rowOffset] = pixel;
              }
          }
      }

      return result;
}
```

Any ideas what the issue may be?

**Comment by brianpopow** [comment, marked as answer]
@LifeguardConcepts could you provide a complete example? What does `BitMatrix matrix` look like?

**Reply by JimBobSquarePants** [reply]
What target framework are you using?

**Reply by LifeguardConcepts** [reply]
The project is targeting .net standard 2.0 and .net 471 on a asp.net core solution.

**Reply by JimBobSquarePants** [reply]
I suspect you are missing binaries that would be installed in the GAC in your CI.

**Reply by JimBobSquarePants** [reply]
>net 471 on a asp.net core solution

ASP.NET Core doesn't support NETFX 4.7.1

**Reply by LifeguardConcepts** [reply]
@brianpopow BitMatrix is essentially just a matrix of either black or white Rgba32 colors which then gets used to determine what color, either a black or white, will be applied on that pixel on the final Image<Rgba32> object.

**Reply by LifeguardConcepts** [reply]
@JimBobSquarePants Not sure what the confusing would be on ASP.NET Core.   Its capable of targeting either .net core or the full .net framework aka net 471 in this particular case.

**Reply by antonfirsov** [reply]
This looks like a build issue caused by inconsistent references. What NuGet packages do you reference?

For troubleshooting, I recommend to create a small ASP.NET core project from scratch with the same targets, referencing the same packages and adding a minimalistic code using both libraries:

```C#
public static void Foo(BitMatrix m) { new Image<Rgba32>(1,1).Dispose(); }
```

If you hit the issue with a minimalistic project, we can reason about your `.csproj`.

**Reply by JimBobSquarePants** [reply]
Depending on the version of ASP.NET Core applications can successfully reference class libraries that target certain versions of .NET standard, _NOT_, .NET Framework. 

I think we’re done here. Tackling people’s build issues is outside of the scope of any unlicensed  support.

**Reply by LifeguardConcepts** [reply]
I agree this is down to a build issue and we can close this out. I was able to override our ci build and it is compiling fine on our release build server.

**Reply by LifeguardConcepts** [reply]
@JimBobSquarePants Just to close this out for future reference .net 471 is the first full .net framework to fully support .net standard 2.0 without additional libraries or binding redirects necessary. Just didn't want someone to see this and think it's not standard compliant for asp.net core targeting standard 2.0

**Reply by JimBobSquarePants** [reply]
@LifeguardConcepts  

> Just didn't want someone to see this and think it's not standard compliant for asp.net core targeting standard 2.0

That's not what you or I said. 

You said
>The project is targeting .net standard 2.0 and .net 471 on a asp.net core solution.

And

>Not sure what the confusing would be on ASP.NET Core. Its capable of targeting either .net core or the full .net framework aka net 471 in this particular case.

That is not how .NET Standard works.

- .NET Framework 4.7.1 implements .NET Standard 2.0 ([_Though you should use NET Framework 4.7.2_](https://docs.microsoft.com/en-us/dotnet/standard/net-standard#net-implementation-support))
- .NET Core 2.0+ implements .NET Standard 2.0

This does not mean that .NET Core 2.0 application should reference NET Framework 4.7.1 libraries. They should reference a .NET Standard <= 2.0 or a .NET Core 2.0 library.
